### PR TITLE
Issue19

### PR DESCRIPTION
**Pull request by technetos**

### CODE REVIEW

**Comment by technetos** [comment]
This PR contains the following functionality to be added to Hatch:
```bash
$ hatch new Test
$ hatch new Test -v 0.1.0
$ hatch new Test -p /some/path/to/Toolbox/C++/libs/project_name
$ hatch new Test --static
$ hatch new Test --binary
$ hatch update
$ hatch update -p /some/path/to/Toolbox/C++/libs/project_name
# and any combination of the above parameters to the given subcommands
```
Calling `hatch new` will generate a new directory with the name of the project name and generates a basic Hatch.yml in that directory

Calling `hatch update` will parse an existing Hatch.yml file, this includes error checking

Both return a new project instance to `bin/hatch.rs` where we can perform other operations on the project if necessary

File: hatch/src/hatch_error/mod.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      HatchError::Io(ref e) => write!(f, "IO error: {}", e),
+      HatchError::Parsing(ref e) => write!(f, "Parsing error: {}", e),
+      HatchError::Null(ref e) => write!(f, "{}", e),
+      HatchError::EmptyConfig(ref e) => write!(f, "Config error: {}", e),
+      HatchError::MissingName(ref e) => write!(f, "Config error: {}", e),
+      HatchError::MissingBuild(ref e) => write!(f, "Config error: {}", e),
+      HatchError::MissingVersion(ref e) => write!(f, "Config error: {}", e),
+    }
+  }
+}
```
Comment (xmclark) [inline hatch/src/hatch_error/mod.rs:58]: There are three methods in this file with match expressions matching the type of error. In each case, we call the same method, e.g. `e.description()` or `write!(f, "message", e)`. **Can this be simpler?** I don't expect these errors to change much, but if we add new errors we have to make sure we add cases in all the match expressions.

**Comment by xmclark** [comment]
You put a lot of effort into this PR. Good job!

**Review by xmclark — DISMISSED** [review]

File: hatch/src/hatch_error/mod.rs (hunk 1 above)
Comment (technetos) [reply hatch/src/hatch_error/mod.rs:58]: All of the errors _statically_ dispatch to the correct call so the size must be known at compile time.  Ive had this same idea and will look into possibly using generics to make this simpler.

Comment (xmclark) [reply hatch/src/hatch_error/mod.rs:58]: 👌

File: hatch/src/hatch_error/mod.rs [hunk 2]
```diff
@@ -28,28 +28,27 @@ pub enum HatchError {
...
   fn cause(&self) -> Option<&Error> {
     match *self {
-      HatchError::Io(ref e) => e.cause(),
-      HatchError::Parsing(ref e) => e.cause(),
-      HatchError::Null(ref e) => e.cause(),
-      HatchError::EmptyConfig(ref e) => e.cause(),
-      HatchError::MissingName(ref e) => e.cause(),
-      HatchError::MissingBuild(ref e) => e.cause(),
-      HatchError::MissingVersion(ref e) => e.cause(),
+      _ => self.cause()
```
Comment (xmclark) [inline hatch/src/hatch_error/mod.rs]: yeah why does this work?

Comment (xmclark) [reply hatch/src/hatch_error/mod.rs]: https://doc.rust-lang.org/std/error/trait.Error.html
All implementations of Error trait have these methods, so it makes sense actually.

File: hatch/src/hatch_error/mod.rs [hunk 3]
```diff
@@ -28,28 +28,27 @@ pub enum HatchError {
...
+//      HatchError::Io(ref e) => e.description(),
+//      HatchError::Parsing(ref e) => e.description(),
+//      HatchError::Null(ref e) => e.description(),
+//      HatchError::EmptyConfig(ref e) => e.description(),
+//      HatchError::MissingName(ref e) => e.description(),
+//      HatchError::MissingBuild(ref e) => e.description(),
+//      HatchError::MissingVersion(ref e) => e.description(),
+//    }
+//  }
+
```
Comment (xmclark) [inline hatch/src/hatch_error/mod.rs]: I would prefer these comments get removed.

**Review by xmclark — CHANGES_REQUESTED** [review]
nice fix!!! Remove the commented code please.

**Comment by technetos** [comment]
Always a good day when I remove lines of code XD

**Review by xmclark — APPROVED** [review]

**Pull request merged**